File: 프로그래머스/2/12978. 배달/배달.py
import heapq
def solution(N, road, K):
    answer = 0
    INF = int(1e9)
    graph = [[] for i in range(N+1)]
    distance = [INF]*(N+1)

    for i in range(0,len(road)):
        a, b, c = road[i][0], road[i][1], road[i][2]
        graph[a].append((b,c))
        graph[b].append((a,c))

    def dijkstra(start):
        q = []
        heapq.heappush(q,(0,start))
        distance[start] = 0
        while q:
            dist, now = heapq.heappop(q)
            if distance[now] < dist:
                continue
            for i in graph[now]:
                cost = dist + i[1]
                if cost < distance[i[0]]:
                    distance[i[0]] = cost
                    heapq.heappush(q,(cost,i[0]))

    dijkstra(1)
    for i in range(1,N+1):
        if distance[i] <= K:
            answer += 1

    return answer

File: 프로그래머스/2/12978. 배달/배달.java
import java.util.*;

class Solution {
    public int solution(int N, int[][] road, int K) {
        int answer = 0;
        
        int[][] graph = new int[N+1][N+1];
        
        for(int[] a : graph){
            Arrays.fill(a, 1000000000);
        }
        
        for(int i = 0; i < graph.length; i++){
            graph[i][i] = 0;
        }
        
        for(int i = 0; i < road.length; i++){
            if (graph[road[i][0]][road[i][1]] > road[i][2]) {
                graph[road[i][0]][road[i][1]] = road[i][2];
                graph[road[i][1]][road[i][0]] = road[i][2];
            }
        }
        
        for(int m = 1; m <= N; m++){
            for(int s = 1; s <= N; s++){
                for(int e = 1; e <= N; e++){
                    int min = graph[s][e] > (graph[s][m] + graph[m][e]) ? graph[s][m] + graph[m][e] : graph[s][e];
                    graph[s][e] = min;
                }
            }
        }
        
        // for(int[] a : graph){
        //     System.out.println(Arrays.toString(a));
        // }
        
        for(int i = 0; i <= N; i++){
            if (graph[1][i] <= K) {
                answer += 1;
            }
        }
        

        return answer;
    }
}